master/buildbot/__init__.py

```diff
@@ -80,6 +80,29 @@ def getVersion(init_file):
     except IOError:
         pass
 
+    # When source is exported via `git archive`, the following line is modified
+    # and placeholders are expanded to the "archived" revision:
+    #
+    #     %ct: committer date, UNIX timestamp
+    #     %d: ref names, like the --decorate option of git-log
+    #
+    # See man gitattributes(5) and git-log(1) (PRETTY FORMATS) for more details.
+    git_archive_id = '$Format:%ct %d$'
```

Comment:
what we can unit test is the extraction method. You can rework that code like this:
```
def getVersionFromArchiveId(git_archive_id):
    if not git_archive_id.startswith('$Format:'):
        # source was modified by git archive, try to parse the version from
        # the value of git_archive_id

        match = re.search(r'tag:\s*v([^,)]+)', git_archive_id)
        if match:
            # archived revision is tagged, use the tag
            return match.group(1)

        # archived revision is not tagged, use the commit date
        tstamp = git_archive_id.strip().split()[0]
        d = datetime.datetime.fromtimestamp(int(tstamp))
        return d.strftime('%Y.%m.%d')
    return None
```
```
def getVersion():
    [...]
    version = getVersionFromArchiveId('$Format:%ct %d$')
    if version is not None:
        return version
```
then the unit test will test results for the different format expected from git-archive (the one without expantion, the one with only timestamp, and the one with timestamp and ref-name)

You can use the same unit test to test the 3 copies of that method (in each module), and put that unit test in buildbot module only. (so that we keep code coverage high, and don't only fix the main one if there is an incompatibility later)